## Credit Risk Prediction

### Table of Content

The layout of this documentation is as follows:

- Business Understanding
- Business Objective
    - Problem Statement
    - Objective
- Data Preparation
    - Import Libraries
    - Import Dataset
    - Features Description
- Data Undertanding
    - Statistical Summary
        - Categorical Features
        - Numerical Features
    - Data Types Information
- Data Preprocessing
    - Create the Target Feature
    - EDA
        - Categorical Features Vs Target Feature
        - Numerical Features Vs Target Feature
        - Total Loan Issued Over Time
        - Calculate Total Loss
    - Data Cleansing
        - Detecting Duplication
        - Detecting Missing Values
- Data Modeling
    - Encoding
    - Feature Selection
    - Handling Imbalanced Data
    - Data Splitting
    - Normalization
    - Machine Learning Techniques
        - Decision Tree
        - Random Forest
        - Logistic Regression
        - K-Nearest Neighbor
        - Gaussian Naive Bayes
        - MLP Classifier (Neural Network)
        - XGBoost Classifier
        - Gradient Boosting Classifier
    - Model Comparison
- Conclusion

## Business Understanding

When the lending company receives a loan application, the company has to decide whether to approve or reject the loans application based on the applicant’s profile (every decision made has a good or bad risk).

- If the applicant is likely to repay the loan, then declining their application will be a business loss to the company. This situation is called a good risk.
- If the applicant is not likely to repay the loan, then approving their application will be a financial loss to the company. This situation is called a bad risk.

The data contains the information about past loans of applicants and whether they labeled as a good risk or not. When a applicant applies for a loan, there are two type of risks, namely:

1. Good risk consists of Fully Paid, Current, and In Grace Period. Applicants with this label are more likely to get their loan approved in the future.
2. Bad Risk consists of Late, Default, and Charged Off. Applicants with this label are unlikely to get their loan approved in the future.

## Business Objective

#### Problem Statement:

Lending loans to ‘bad risk’ applicants is the largest source of financial loss. Credit loss is the amount of money lost by the lender when the applicant refuses to pay or runs away with the money owed.

#### Objectives:

1. Identify patterns that indicate if a person is unlikely to repay the loan or labeled as a bad risk so that it can be used to take action such as rejecting the loan, reducing the amount of loan, lending at a higher interest rate, etc.
2. Implement machine learning algorithms to build predictive model so that the company can automatically predict whether the loan application submitted by the applicant will labeled as a bad risk or not. With this, the company can make a decision to approve or reject the loan application.

## Data Preparation

### Import Libraries

In [1]:
# import libraries
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import warnings # ignores any warning
warnings.filterwarnings('ignore')

In [6]:
pd.set_option('display.max_columns',None)
pd.set_option('display.max_rows',None)
pd.set_option('float_format', '{:f}'.format)

### Import Dataset

In [7]:
# import dataset
df = pd.read_csv('loan_data.csv')
print('This dataset has %d rows and %d columns.\n' % df.shape)
df.head()

This dataset has 466285 rows and 75 columns.



Unnamed: 0       id  member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
0           0  1077501    1296599       5000         5000      4975.000000   
1           1  1077430    1314167       2500         2500      2500.000000   
2           2  1077175    1313524       2400         2400      2400.000000   
3           3  1076863    1277178      10000        10000     10000.000000   
4           4  1075358    1311748       3000         3000      3000.000000   

         term  int_rate  installment grade sub_grade  \
0   36 months 10.650000   162.870000     B        B2   
1   60 months 15.270000    59.830000     C        C4   
2   36 months 15.960000    84.330000     C        C5   
3   36 months 13.490000   339.310000     C        C1   
4   60 months 12.690000    67.790000     B        B5   

                  emp_title emp_length home_ownership   annual_inc  \
0                       NaN  10+ years           RENT 24000.000000   
1                     Ryder   < 1 year           RENT 30000.000000   
2                       NaN  10+ years           RENT 12252.000000   
3       AIR RESOURCES BOARD  10+ years           RENT 49200.000000   
4  University Medical Group     1 year           RENT 80000.000000   

  verification_status issue_d  loan_status pymnt_plan  \
0            Verified  Dec-11   Fully Paid          n   
1     Source Verified  Dec-11  Charged Off          n   
2        Not Verified  Dec-11   Fully Paid          n   
3     Source Verified  Dec-11   Fully Paid          n   
4     Source Verified  Dec-11      Current          n   

                                                 url  \
0  https://www.lendingclub.com/browse/loanDetail....   
1  https://www.lendingclub.com/browse/loanDetail....   
2  https://www.lendingclub.com/browse/loanDetail....   
3  https://www.lendingclub.com/browse/loanDetail....   
4  https://www.lendingclub.com/browse/loanDetail....   

                                                desc         purpose  \
0    Borrower added on 12/22/11 > I need to upgra...     credit_card   
1    Borrower added on 12/22/11 > I plan to use t...             car   
2                                                NaN  small_business   
3    Borrower added on 12/21/11 > to pay for prop...           other   
4    Borrower added on 12/21/11 > I plan on combi...           other   

                  title zip_code addr_state       dti  delinq_2yrs  \
0              Computer    860xx         AZ 27.650000     0.000000   
1                  bike    309xx         GA  1.000000     0.000000   
2  real estate business    606xx         IL  8.720000     0.000000   
3              personel    917xx         CA 20.000000     0.000000   
4              Personal    972xx         OR 17.940000     0.000000   

  earliest_cr_line  inq_last_6mths  mths_since_last_delinq  \
0           Jan-85        1.000000                     NaN   
1           Apr-99        5.000000                     NaN   
2           Nov-01        2.000000                     NaN   
3           Feb-96        1.000000               35.000000   
4           Jan-96        0.000000               38.000000   

   mths_since_last_record  open_acc  pub_rec  revol_bal  revol_util  \
0                     NaN  3.000000 0.000000      13648   83.700000   
1                     NaN  3.000000 0.000000       1687    9.400000   
2                     NaN  2.000000 0.000000       2956   98.500000   
3                     NaN 10.000000 0.000000       5598   21.000000   
4                     NaN 15.000000 0.000000      27783   53.900000   

   total_acc initial_list_status  out_prncp  out_prncp_inv  total_pymnt  \
0   9.000000                   f   0.000000       0.000000  5861.071414   
1   4.000000                   f   0.000000       0.000000  1008.710000   
2  10.000000                   f   0.000000       0.000000  3003.653644   
3  37.000000                   f   0.000000       0.000000 12226.302210   
4  38.000000                   f 766.900000     766.900000  3242.17

There are 75 features and 466,285 rows which contains the information about past loans of applicants.

The `loan_status` feature defines the past loan status of applicants, which is Current, Fully Paid, Charged Off, Late (16-30 days), Late (31-120 days), In Grace Period, and Default. This feature will be the target feature for credit risk prediction analysis.

## Data Understanding

### Statistical Summary

### Numerical Features

In [9]:
num_features = df.select_dtypes(include=['int64','float64'])
print('The number of numerical features is {}'.format(num_features.shape[1]))

The number of numerical features is 53


In [11]:
num_features.describe().transpose()

count            mean             std  \
Unnamed: 0                  466285.000000   233142.000000   134605.029472   
id                          466285.000000 13079729.906182 10893711.624551   
member_id                   466285.000000 14597660.622827 11682367.441450   
loan_amnt                   466285.000000    14317.277577     8286.509164   
funded_amnt                 466285.000000    14291.801044     8274.371300   
funded_amnt_inv             466285.000000    14222.329888     8297.637788   
int_rate                    466285.000000       13.829236        4.357587   
installment                 466285.000000      432.061201      243.485550   
annual_inc                  466281.000000    73277.381470    54963.568654   
dti                         466285.000000       17.218758        7.851121   
delinq_2yrs                 466256.000000        0.284678        0.797365   
inq_last_6mths              466256.000000        0.804745        1.091598   
mths_since_last_delinq      215934.000000       34.104430       21.778487   
mths_since_last_record       62638.000000       74.306012       30.357653   
open_acc                    466256.000000       11.187069        4.987526   
pub_rec                     466256.000000        0.160564        0.510863   
revol_bal                   466285.000000    16230.203487    20676.245152   
revol_util                  465945.000000       56.176947       23.732628   
total_acc                   466256.000000       25.064430       11.600141   
out_prncp                   466285.000000     4410.062342     6355.078769   
out_prncp_inv               466285.000000     4408.452258     6353.198001   
total_pymnt                 466285.000000    11540.686220     8265.627112   
total_pymnt_inv             466285.000000    11469.892747     8254.157579   
total_rec_prncp             466285.000000     8866.014657     7031.687997   
total_rec_int               466285.000000     2588.677225     2483.809661   
total_rec_late_fee          466285.000000        0.650129        5.265730   
recoveries                  466285.000000       85.344211      552.216084   
collection_recovery_fee     466285.000000        8.961534       85.491437   
last_pymnt_amnt             466285.000000     3123.913796     5554.737393   
collections_12_mths_ex_med  466140.000000        0.009085        0.108648   
mths_since_last_major_derog  98974.000000       42.852547       21.662591   
policy_code                 466285.000000        1.000000        0.000000   
annual_inc_joint                 0.000000             NaN             NaN   
dti_joint                        0.000000             NaN             NaN   
verification_status_joint        0.000000             NaN             NaN   
acc_now_delinq              466256.000000        0.004002        0.068637   
tot_coll_amt                396009.000000      191.913517    14630.214605   
tot_cur_bal                 396009.000000   138801.713385   152114.663494   
open_acc_6m                      0.000000             NaN             NaN   
open_il_6m                       0.000000             NaN             NaN   
open_il_12m                      0.000000             NaN             NaN   
open_il_24m                      0.000000             NaN             NaN   
mths_since_rcnt_il               0.000000             NaN             NaN   
total_bal_il                     0.000000             NaN             NaN   
il_util                          0.000000             NaN             NaN   
open_rv_12m                      0.000000             NaN             NaN   
open_rv_24m                      0.000000             NaN             NaN   
max_bal_bc                       0.000000             NaN             NaN   
all_util                         0.000000             NaN             NaN   
total_rev_hi_lim            396009.000000    30379.087771    37247.129571   
inq_fi                           0.000000             NaN             NaN   
total_cu_tl                     

- `Unnamed: 0`, `id`, `member_id` are unique for each row. This features will be removed as because it is not needed for analysis.
- policy_code has only one unique value. This feature will be removed because no information can be obtained from the this feature.
- There are 17 features that have missing value. This features will be dealt with in the data clansing section.

In [12]:
# drop unnecessary features
df.drop(['Unnamed: 0','member_id','policy_code'],inplace=True,axis=1)

### Categorical Features

In [13]:
cat_features = df.select_dtypes(include=['object'])
print('The number of categorical feature is {}'.format(cat_features.shape[1]))

The number of categorical feature is 22


In [14]:
cat_features.describe().transpose()

count  unique  \
term                 466285       2   
grade                466285       7   
sub_grade            466285      35   
emp_title            438697  205475   
emp_length           445277      11   
home_ownership       466285       6   
verification_status  466285       3   
issue_d              466285      91   
loan_status          466285       9   
pymnt_plan           466285       2   
url                  466285  466285   
desc                 125983  124436   
purpose              466285      14   
title                466265   63099   
zip_code             466285     888   
addr_state           466285      50   
earliest_cr_line     466256     664   
initial_list_status  466285       2   
last_pymnt_d         465909      98   
next_pymnt_d         239071     100   
last_credit_pull_d   466243     103   
application_type     466285       1   

                                                                   top    freq  
term                                                         36 months  337953  
grade                                                                B  136929  
sub_grade                                                           B3   31686  
emp_title                                                      Teacher    5399  
emp_length                                                   10+ years  150049  
home_ownership                                                MORTGAGE  235875  
verification_status                                           Verified  168055  
issue_d                                                         Oct-14   38782  
loan_status                                                    Current  224226  
pymnt_plan                                                           n  466276  
url                  https://www.lendingclub.com/browse/loanDetail....       1  
desc                                                                       234  
purpose                                             debt_consolidation  274195  
title                                               Debt consolidation  164075  
zip_code                                                         945xx    5304  
addr_state                                                          CA   71450  
earliest_cr_line                                                Oct-00    3674  
initial_list_status                                                  f  303005  
last_pymnt_d                                                    Jan-16  179620  
next_pymnt_d                                                    Feb-16  208393  
last_credit_pull_d                                              Jan-16  327699  
application_type                                            INDIVIDUAL  466285

- `emp_title`, `url` ,`desc`, `title`, `zip_code`, and `addr_state` have many unique values. This features will be removed.
- `application_type` has only one unique value. This feature will be removed because no information can be obtained from the this feature.

In [15]:
# drop unnecessary features
df.drop(['emp_title','url','desc','title','zip_code','addr_state'],inplace=True,axis=1)

### Data Types Information

In [16]:
print('Data type before correction:\n')
df.info()

Data type before correction:

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 466285 entries, 0 to 466284
Data columns (total 66 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   id                           466285 non-null  int64  
 1   loan_amnt                    466285 non-null  int64  
 2   funded_amnt                  466285 non-null  int64  
 3   funded_amnt_inv              466285 non-null  float64
 4   term                         466285 non-null  object 
 5   int_rate                     466285 non-null  float64
 6   installment                  466285 non-null  float64
 7   grade                        466285 non-null  object 
 8   sub_grade                    466285 non-null  object 
 9   emp_length                   445277 non-null  object 
 10  home_ownership               466285 non-null  object 
 11  annual_inc                   466281 non-null  float64
 12  verification_status         

- `issue_d`, `earliest_cr_line`, `last_pymnt_d`, `next_pymnt_d`, and `last_credit_pull_d` will be converted to datetime forma

In [17]:
def date_time(dt):
    if dt.year > 2016:
        dt = dt.replace(year = dt.year - 100)
    return dt

In [18]:
# convert string to datetime
from datetime import datetime
df['earliest_cr_line'] = pd.to_datetime(df['earliest_cr_line'], format='%b-%y')
df['earliest_cr_line'] = df['earliest_cr_line'].apply(lambda x: date_time(x))
df['issue_d'] = pd.to_datetime(df['issue_d'], format='%b-%y')
df['next_pymnt_d'] = pd.to_datetime(df['next_pymnt_d'], format='%b-%y')
df['last_pymnt_d'] = pd.to_datetime(df['last_pymnt_d'], format='%b-%y')
df['last_credit_pull_d'] = pd.to_datetime(df['last_credit_pull_d'], format='%b-%y')
df[['earliest_cr_line','issue_d','next_pymnt_d','last_pymnt_d','last_credit_pull_d']].head()

earliest_cr_line    issue_d next_pymnt_d last_pymnt_d last_credit_pull_d
0       1985-01-01 2011-12-01          NaT   2015-01-01         2016-01-01
1       1999-04-01 2011-12-01          NaT   2013-04-01         2013-09-01
2       2001-11-01 2011-12-01          NaT   2014-06-01         2016-01-01
3       1996-02-01 2011-12-01          NaT   2015-01-01         2015-01-01
4       1996-01-01 2011-12-01   2016-02-01   2016-01-01         2016-01-01

In [19]:
print('Data type after correction\n')
df.info()

Data type after correction

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 466285 entries, 0 to 466284
Data columns (total 66 columns):
 #   Column                       Non-Null Count   Dtype         
---  ------                       --------------   -----         
 0   id                           466285 non-null  int64         
 1   loan_amnt                    466285 non-null  int64         
 2   funded_amnt                  466285 non-null  int64         
 3   funded_amnt_inv              466285 non-null  float64       
 4   term                         466285 non-null  object        
 5   int_rate                     466285 non-null  float64       
 6   installment                  466285 non-null  float64       
 7   grade                        466285 non-null  object        
 8   sub_grade                    466285 non-null  object        
 9   emp_length                   445277 non-null  object        
 10  home_ownership               466285 non-null  object        
 11

## Data Preprocessing

### Create Target Feature

### Loan Status Description

- **Fully paid** means the loan has been fully repaid, either at the end of the loan term or earlier because of prepayment.


- **Current** means the applicants is making payments on time.


- **In grace period** is a set number of days after the due date during which payment may be made by the applicants without penalty. The exact number of days is determined by each lending company. During this period no late fees are charged, and the delay cannot result in default or cancellation of the loan or contract. In most cases, payment after the due date but during the grace period does not cause a black mark to be added to the applicants’s credit report.


- **Late means** the applicants failed to make a timely payment within the grace period. In this case, grace period days will be counted as late days. Late payments can hurt applicants credit scores, although the impact will depend on applicants overall credit profile and how far behind applicants fall on their payments. Late payments are further broken down by the number of late days.


    - 16-30 days late
    - 31-120 days late


- **Default** means the loan agreement has been terminated by the lending company, and the outstanding balance of the loan is due and payable. Default occurs when a applicants fails to pay back a debt according to the initial arrangement. Defaulting on a loan happens when repayments aren't made for a certain period of time. When a loan defaults, it is sent to a debt collection agency whose job is to contact the applicant and receive the unpaid funds. Defaulting will drastically reduce your credit score, impact your ability to receive future credit, and can lead to the seizure of personal property.


- **Charged Off** is if applicants been delinquent on their credit card or loan payments for several months. This occurs when the creditor has given up on collecting the money owed and has decided to categorize the debt as bad debt, meaning it is a loss for the company. This is categorized as bad debt. Bad debt means there is no longer a reasonable expectation of further payments on this loan. The bailiff or court has declared no funds can be recovered.

Based on the explanation above, I will create a new feature, namely risk status which consists of 2 classes.

1. **Good Risk** class consists of Fully Paid, Current, and In Grace Period.

    - This selection is based on the rule that if the applicants is in these period, then the applicant's credit score will not be affected or does not cause a black mark to be added to the applicant's credit report.
    - This class means that if the applicant applies for a loan again and is likely to repay the loan, then not approving the loan results in a loss of business to the company.
    

2. **Bad Risk** class consist of Default, Late, and Charged Off.

    - It is because if the applicants is in this period, it will affect their credit score so that it can impact their ability to receive future credit.
    - This class means that if the applicant applies for a loan again and is not likely to repay the loan, then approving the loan may lead to a financial loss for the company.

In [20]:
# create target feature
df['risk'] = np.where((df['loan_status'] == 'Charged Off') |
                     (df['loan_status'] == 'Default') |
                     (df['loan_status'] == 'Late (31-120 days)') |
                     (df['loan_status'] == 'Late (16-30 days)') |
                     (df['loan_status'] == 'Does not meet the credit policy. Status: Charged Off'), 'Bad Risk', 'Good Risk')

## Exploratory Data Analysis (EDA)

### Univariate: The Number of Applicants by Loan Status

In [31]:
# table
loan_grp = df.groupby('loan_status').size().reset_index()
loan_grp.columns = ['target','total']
loan_grp['%'] = round(loan_grp['total']*100/sum(loan_grp['total']),2)
loan_grp.sort_values(by='%', ascending=False).style.background_gradient(cmap='Blues')

In [29]:
loan_grp['total']


0     42475
1    224226
2       832
3       761
4      1988
5    184739
6      3146
7      1218
8      6900
Name: total, dtype: int64